Updated Working-with-ResultAsync (#1)
lauhon/neverthrow-wiki-fork: Working-with-ResultAsync.md

> Result Async can be used to work with Asynchronous Results in an Error safe way.

Reviewer: Suggestion / nitpick:

```suggestion
Result Async can be used to work with Asynchronous Results in a typesafe way.
```

> Similarly to `Result<T,E>`, which contains the Result of a synchronous operation that may have failed, you can handle Promises with `ResultAsync<T,E>`.

Reviewer: ```suggestion
Similarly to `Result<T,E>`, which contains the Result of a synchronous operation that may have failed, you can represent the result of async operations with `ResultAsync<T,E>`.

`ResultAsync<T, E>` has the exact same API as `Promise<T>`, but with many more methods, and also a dedicated type (`E`) for the error you expect on your async operation.
```

and how it should handle the asynchronous action.

### Example

Below are three functions. Each of them handling the same error in a different way

```typescript

const callbackUser = () => {
    const userPromise = prisma.user.findUnique(...)

    userPromise
    .then(user => console.log("Found my user", user))
    .catch(err => console.error("Error while reading user", err))
}

const asyncAwaiter = async () => {
    try {
        const user = await prisma.user.findUnique(...)
        console.log("Found my user", user)
    } catch (err) {
        console.error("Error while reading user", err)
    }
}

const neverthrower = () => {
    const userPromise = prisma.user.findUnique(...)

    const result = fromPromise(userPromise, originalError => ({originalError, message: "Error while reading user"}))

    result.match(
        user => console.log("Found my user", user),
        err => console.log(err.message, err.originalError)
    )
}
```

Notice that typically you would not read the result of ResultAsync immediately after wrapping it, as it does not much other than bloating up your function a bit more.

> The gain of using this paradigm really comes to the surface when you start to chaining different `Result` and `ResultAsnyc` together with the `.map` and `.andThen` functions. This enables you to unwrap (with `match`, or `map` & `mapErr`) the Results at a centralized point to outsource the responsibility to react to Errors and Successes.

Reviewer: ```suggestion
The benefit of using this paradigm really comes to the surface when you start chaining different `Result` and `ResultAsnyc` together with the `.map` and `.andThen` functions. This enables you to unwrap (with `match`, or `map` & `mapErr`) the Results at a centralized point to outsource the responsibility to react to Errors and Successes.
```
